### ISSUE TITLE
Add subsampling option

### ISSUE THREAD

**Issue by huyduongtu** [koral--/android-gif-drawable#239]
Firstly, thanks for your library. It works well and save my life a lot
In your lib, you create a Bitmap in java heap with size based on the width and height of each frame. Foe each frame, you set the corresponding color pixel of the bitmap in native.
But If the gif size is so big (width and height of each frame are big also) and the imageview is small. I'm afraid of creating the bitmap can causes OutOfMemoryError in low-end devices.
How can I scale down the bitmap size?
I mean, I can pass the actual width, height of the ImageView, then the library can scale down the frame size base on those argument. This is similar to BitmapFactory.Options.inSampleSize.
Thanks in advance.

**Comment by koral--** [comment]
Currently there is no way reduce size of bitmaps used as frame buffers in `GifDrawable` from code. There are only indirect ways to reduce overall memory usage in given moment of time eg. by reusing existing `GifDrawable` instances instead of creating them from scratch.
However I'll add resampling option ASAP.

**Comment by huyduongtu** [comment]
Thanks for your quick response. Could you tell the way to reuse existing GifDrawable?
And when subsampling option is available?

**Comment by koral--** [comment]
Reusing existing `GifDrawable` is similar to [BitmapFactory.Options.inBitmap](http://developer.android.com/reference/android/graphics/BitmapFactory.Options.html#inBitmap). You have to pass existing drawable to `GifDrawableBuilder`. I've added that to [sample project](https://github.com/koral--/android-gif-drawable-sample/blob/master/sample/src/main/java/pl/droidsonroids/gif/sample/GifSourcesFragment.java#L114).
Another workaround is to use `GifTextureView`, in such case no `Bitmap`s are needed.
I can't promise exact time when subsampling will be added.

**Comment by huyduongtu** [comment]
You said "Sets drawable to be reused when creating new one. Currently it works only on  KITKAT and newer, on older API levels call has no effect.". Older version can not use this way.
About GifTextureView, I don't understand how it works. Can you clarify it for me?
The secret is 
```
 final Surface surface = new Surface(surfaceTexture);
 try {
      mGifInfoHandle.bindSurface(surface, mSavedState, gifTextureView.isOpaque());
 } finally {
      surface.release();
 }
```
Is it right?

**Comment by koral--** [comment]
Note about Kitkat in javadoc is outdated, this issue reminded me to remove it and it is not present in the latest snapshot. Drawable reusing is working in API levels.

For `GifTextureView` you have to use [#setInputSource](http://koral--.github.io/android-gif-drawable/javadoc/pl/droidsonroids/gif/GifTextureView.html#setInputSource-pl.droidsonroids.gif.InputSource-). Keep in mind that TextureView is available since API level 14 and requires hardware accelerated window.

**Comment by huyduongtu** [comment]
Thanks. Hope to see the next subsampling feature soon.

**Comment by huyduongtu** [comment]
One more question. I'm implementing a cache that keeps GifDrawable in memory. I wonder how much space does GifDrawable take in memory? It is much?
In my opinion, GifDrawable just has only one bitmap for the current frame, the rest is in native memory. So it is ok to keep them in memory? Right?

**Comment by koral--** [comment]
There is always one `Bitmap` in terms of `android.graphics.Bitmap`. If GIF uses "Restore to previous" disposal method then additionally memory of size equal to bitmap (4 bytes per pixel) is allocated in native code.

**Comment by huyduongtu** [comment]
So Does your lib use "Restore to previous" disposal method? Because native memory is stilled counted in limited memory used for an Android application. I need to know to consider how much space for the GifDrawable cache.

**Comment by koral--** [comment]
Not sure if I understand correctly, but try to answer.
Library does not "use" "Restore to previous". Only creator of a GIF file can use it. If GIF is animated then each frame descriptor contains information how it should be disposed. There are 3 methods available: http://www.imagemagick.org/Usage/anim_basics/#dispose
Library supports all of them appropriately.

**Comment by huyduongtu** [comment]
So the most memory space taken by an instance GifDrawable is a Bitmap (in heap memory)?The rest is the information of the Gif file?

**Comment by koral--** [comment]
It is true that that most of the memory is consumed by `Bitmap` in `GifDrawable` (width * height * 4 bytes + overhead of Bitmap internal data) but in the rest apart from information about GIF file there are 2 things:
- aforementioned additional framebuffer for "Restore to previous" disposal (allocated only if GIF file uses it)
- buffer for raster, allocated here: https://github.com/koral--/android-gif-drawable/blob/master/src/main/jni/open_close.c#L62, size is width * height (1 byte per pixel) assuming that your compiler uses `sizeof(char) == 1`, which is most common. Raster contains not colors (RGBA values) but indices of colour in palette thus 1 byte is enough.

**Comment by huyduongtu** [comment]
Got it. Thanks.

**Comment by huyduongtu** [comment]
Got it. Thanks.

**Comment by manoj99** [comment]
your library is very helpful it is load GIF with low efficiency.is there any method does your libaray can build gif from combine bitmap?

**Comment by koral--** [comment]
@manoj99 unfortunately not yet, but such option may be added in the future.

**Comment by koral--** [comment]
Subsampling draft:
* size of the canvas, frame and offsets should be seen scaled everywhere except `DDGifSlurp`, so drawing logic should work OOTB
* dimensions scaling will occur in `DDGifSlurp` just after calling `DGifGetImageDesc`
* raster size will be unchanged but after calling `DGifGetLine` it will be subsampled (using the same memory block)
* it must be guaranteed that canvas size will not be smaller than 1x1 px

**Comment by koral--** [comment]
Subsampling option is now available on dev branch and on maven snapshot repository.

**Comment by huyduongtu** [comment]
Thanks. Is it stable?

**Comment by koral--** [comment]
Yeah, it should be.

**Comment by koral--** [comment]
It look like everything is ready, including javadocs and subsampling is going to be included in v 1.1.14.

**Comment by huyduongtu** [comment]
Thanks. Wait for it is available on the master branch.
